### PR TITLE
feat: [CDS-77592]: get support for remote services 

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
Support GET for remote services with fallback branch 
Provides a custom method with fetch from cache and fallback branch options 
Adds common methods to a utility class 

Add a common method to populate entity details from both metadata and data from SCM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52283)

### CODE REVIEW

File: 127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java [hunk 1]
```diff
@@ -227,4 +237,85 @@ private void addGitParamsToServiceEntity(ServiceEntity serviceEntity, GitEntityI
...
+        .is(orgIdentifier)
+        .and(ServiceEntityKeys.projectIdentifier)
+        .is(projectIdentifier)
+        .and(ServiceEntityKeys.identifier)
+        .is(serviceIdentifier)
+        .and(ServiceEntityKeys.deleted)
+        .is(deleted);
+  }
+
+  ServiceEntity fetchRemoteEntity(String accountId, String orgIdentifier, String projectIdentifier,
```
Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java:283]: should be private ?

**Review by yogesh-chauhan — DISMISSED** [review]

**Review by tchaurasiya — DISMISSED** [review]

**Comment by hingerabhinav** [comment]
trigger uts

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 4 of 7 files at r1, 1 of 2 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52283)* status: 13 file(s) to review. 0 of 2 approvals obtained.

**Comment by hingerabhinav** [comment]
trigger uts

**Comment by hingerabhinav** [comment]
trigger uts

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 4 of 4 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52283)* status: 15 file(s) to review. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by adivishy1** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java` line 253 at r7](https://reviewable.io/reviews/harness/harness-core/52283#-NdUOWFD5HUJXVhl0j55:-NdUOWFD5HUJXVhl0j56:b-xq163k) ([raw file](https://github.com/harness/harness-core/blob/fe17602fb24fecea1bb4dbe4ae52e85ead11a7e2/127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java#L253)):*
> ```Java
>       return Optional.empty();
>     }
> ```

are integrating fetch metadata only flow in another pr?
for few cases, for remote entities also we might just need the metadata and not really fetch from git, 
by doing so we can save the response time

**Comment by adivishy1** [comment]
Have we tested out all the flows, for a gitx project:
1. inline entity
2. remote entity [default and for non default if load from fallback works]
3. nested remote entities [same repo, same branch else different repo different branch]

**Comment by hingerabhinav** [comment]
> _[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java` line 253 at r7](https://reviewable.io/reviews/harness/harness-core/52283#-NdUOWFD5HUJXVhl0j55:-NdUOWFD5HUJXVhl0j56:b-xq163k) ([raw file](https://github.com/harness/harness-core/blob/fe17602fb24fecea1bb4dbe4ae52e85ead11a7e2/127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java#L253)):_
> 
> > ```java
> >       return Optional.empty();
> >     }
> > ```
> 
> are integrating fetch metadata only flow in another pr? for few cases, for remote entities also we might just need the metadata and not really fetch from git, by doing so we can save the response time

Yes @adivishy1 we are created a separate flow/API where we need just the metadata. This way seemed the best because the caller does not need to be aware of the internal working of SCM context

**Review by adivishy1 — APPROVED** [review]
> Reviewed 4 of 7 files at r1, 1 of 2 files at r3, 1 of 8 files at r4, 4 of 4 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52283)* status: :shipit: complete! ✓

**Comment by hingerabhinav** [comment]
trigger review

**Pull request merged**